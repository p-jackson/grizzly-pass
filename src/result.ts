export interface Result<T, E> {
  map<T2>(f: (t: T) => T2): Result<T2, E>;
  mapErr<E2>(f: (e: E) => E2): Result<T, E2>;
  isOk(): boolean;
  isErr(): boolean;
  andThen<T2, E2>(f: (t: T) => Result<T2, E2>): Result<T2, E | E2>;
  orElse<T2, E2>(f: (e: E) => Result<T2, E2>): Result<T | T2, E2>;
  unsafeUnwrap(): T;
}

export class Ok<T, E> implements Result<T, E> {
  value: T;

  constructor(value: T) {
    this.value = value;
  }

  map<T2>(f: (t: T) => T2): Result<T2, E> {
    return new Ok<T2, E>(f(this.value));
  }

  mapErr<E2>(f: (e: E) => E2): Result<T, E2> {
    return new Ok<T, E2>(this.value);
  }

  isOk() {
    return true;
  }

  isErr() {
    return false;
  }

  andThen<T2, E2>(f: (t: T) => Result<T2, E2>): Result<T2, E | E2> {
    return f(this.value);
  }

  orElse<T2, E2>(f: (e: E) => Result<T2, E2>): Result<T | T2, E2> {
    return new Ok<T, E2>(this.value);
  }

  unsafeUnwrap(): T {
    return this.value;
  }
}

export class Err<T, E> implements Result<T, E> {
  error: E;

  constructor(error: E) {
    this.error = error;
  }

  map<T2>(f: (t: T) => T2): Result<T2, E> {
    return new Err<T2, E>(this.error);
  }

  mapErr<E2>(f: (e: E) => E2): Result<T, E2> {
    return new Err<T, E2>(f(this.error));
  }

  isOk() {
    return false;
  }

  isErr() {
    return true;
  }

  andThen<T2, E2>(f: (t: T) => Result<T2, E2>): Result<T2, E | E2> {
    return new Err<T2, E>(this.error);
  }

  orElse<T2, E2>(f: (e: E) => Result<T2, E2>): Result<T | T2, E2> {
    return f(this.error);
  }

  unsafeUnwrap(): T {
    if (typeof this.error === "string" || this.error.toString)
      throw new Error(`Unwrapped result with err: ${this.error}`);
    else throw new Error(`Unwrapped result with err`);
  }
}

export const ok = <T>(t: T) => new Ok<T, any>(t);

export const err = <E>(e: E) => new Err<any, E>(e);

export function fromList<T, E>(
  listOfResults: Result<T, E>[]
): Result<T[], E[]> {
  const errs = listOfResults.reduce((memo, r) => {
    if (_isErr(r)) return [...memo, r.error];
    else return memo;
  }, []);
  if (errs.length) return err(errs);
  else return ok(listOfResults.map(r => r.unsafeUnwrap()));
}

function _isErr<T, E>(result: Result<T, E>): result is Err<T, E> {
  return result.isErr();
}
