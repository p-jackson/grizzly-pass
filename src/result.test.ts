import * as assert from "assert";
import { err, ok, fromList } from "./result";

it("returns the value if an ok result is unsafeUnwrap'd", () => {
  const value = ok(113).unsafeUnwrap();
  expect(value).toBe(113);
});

it("throws if an error Result<_, string> is unsafeUnwrap'd", () => {
  expect(() => err("bad").unsafeUnwrap()).toThrowErrorMatchingSnapshot();
});

it("throws if an error Result<_, []> is unsafeUnwrap'd", () => {
  expect(() =>
    err(["bad", "bad"]).unsafeUnwrap(),
  ).toThrowErrorMatchingSnapshot();
});

it("throws if an error Result<_, T> is unsafeUnwrap'd (where T doesn't have a toString)", () => {
  const obj = Object.create(null);
  expect(() => err(obj).unsafeUnwrap()).toThrowErrorMatchingSnapshot();
});

it("leaves an ok result unchanged when mapErr is called", () => {
  const result = ok(113).mapErr(() => assert.fail());
  expect(result).toEqual(ok(113));
});

it("changes an err result when mapErr is called", () => {
  const result = err(113).mapErr((i) => i + 1);
  expect(result).toEqual(err(114));
});

it("changes an ok result when map is called", () => {
  const result = ok(113).map((i) => i + 1);
  expect(result).toEqual(ok(114));
});

it("leaves an err result unchanged when map is called", () => {
  const result = err(113).map(() => assert.fail());
  expect(result).toEqual(err(113));
});

it("uses the returned result when andThen is called", () => {
  const result = ok(113).andThen(() => err("to error"));
  expect(result).toEqual(err("to error"));
});

it("leaves an err result unchanged when andThen is called", () => {
  const result = err(113).andThen(() => assert.fail());
  expect(result).toEqual(err(113));
});

it("uses the returned result when orElse is called", () => {
  const result = err(113).orElse((i) => err(i + 1));
  expect(result).toEqual(err(114));
});

it("leaves an ok result unchanged when orElse is called", () => {
  const result = ok(113).orElse(() => assert.fail());
  expect(result).toEqual(ok(113));
});

it("returns true from isOk when result is ok", () => {
  expect(ok(undefined).isOk()).toBe(true);
});

it("returns false from isOk when result is err", () => {
  expect(err(undefined).isOk()).toBe(false);
});

it("returns true from isErr when result is err", () => {
  expect(err(undefined).isErr()).toBe(true);
});

it("returns false from isErr when result is ok", () => {
  expect(ok(undefined).isErr()).toBe(false);
});

describe("fromList", () => {
  it("returns an ok result if all list elements are ok results", () => {
    expect(fromList([ok(1), ok(2), ok(3)])).toEqual(ok([1, 2, 3]));
  });

  it("results an err result if all list elements are err results", () => {
    expect(fromList([err(1), err(2)])).toEqual(err([1, 2]));
  });

  it("returns an err result if any list element is an err result", () => {
    expect(fromList([ok(1), err(2)])).toEqual(err([2]));
  });
});
