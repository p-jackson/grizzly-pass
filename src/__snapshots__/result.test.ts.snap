// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if an error Result<_, []> is unsafeUnwrap'd 1`] = `"Unwrapped result with err: bad,bad"`;

exports[`throws if an error Result<_, T> is unsafeUnwrap'd (where T doesn't have a toString) 1`] = `"Unwrapped result with err"`;

exports[`throws if an error Result<_, string> is unsafeUnwrap'd 1`] = `"Unwrapped result with err: bad"`;
